build: eliminate `i18n_tool.py`

cfm opened the pull request:
## Status

Ready for review

## Description of Changes

Closes #6917 by removing `i18n_tool.py` in favor of:
- `make extract-strings`, which wraps `pybabel extract`;
- `make check-{desktop-files,strings}`, which enforces `make extract-strings` in CI;
- `pybabel compile`;
- `make supported-locales`, which consumes `i18n.json` and can be further processed via `jq`;
- #6953;
- freedomofpress/securedrop-docs#503; and
- freedomofpress/securedrop-dev-docs#86.

## Testing

Because this pull request offloads much of our existing localization workflow onto Weblate, very little self-contained, local testing is possible or worthwhile.  Namely:

- Inspect packages for differences between `develop` and here, with a fresh build image each time:
    1. `(docker|podman) image prune -a --force && make build-debs` on `develop` and save them (e.g., `cp -Rp build build@develop`).
    2. `(docker|podman) image prune -a --force && make build-debs` on this branch.
    3. [x] `diffoscope build@develop/focal/securedrop-app-code*.deb build/focal/securedrop-app-code*.deb` shows only:
        - [x] diffs from this branch
        - [x] metadata changes in `.mo` files

Instead...

### Perform a mock release and deployment via `securedrop-i18n-sandbox`

A mock Localization Manager (@cfm) and Release Manager (@nathandyer) should pair their way through (approximately) the following:

1. [x] Coordinate with Infrastructure so that: freedomofpress/infrastructure#4393
    - [x] `weblate-sandbox.securedrop.org` uses...
    - [x] GitHub user @weblate-sandbox-fpf to open pull requests in...
    - [x] `freedomofpress/securedrop-i18n-sandbox`.
2. [x] **At `weblate-sandbox.securedrop.org`,** configure `securedrop/securedrop` and `securedrop-desktop` (from `freedomofpress/securedrop-i18n-sandbox`) like `securedrop/securedrop-client` (from `freedomofpress/securedrop-client-i18n-sandbox`).  **Add-ons:**
    - [x] "Update PO files to match POT"
         - [x] Keep previous msgids of translated strings
         - [x] Remove locations of translated strings
    - [x] "Customize gettext output"
        - [x] Long lines wrapping: No line wrapping
    - [x] "Squash Git commits"
        - [x] Commit squashing: Per language
        - [x] Append trailers to squash commit
    - [x] `desktop` only: Generate MO files
4. [x] Push this branch to `securedrop-i18n-sandbox@i18n` (as though it were already merged into `securedrop@develop`).
5. [x]  **In `securedrop-i18n-sandbox`,** make, commit, and push a change to a source string...
    - [x] in `securedrop/`
    - [x] in `install_files/ansible-base/roles/tails-config/templates/locale/`
6. **At `weblate-sandbox.securedrop.org`,** make, commit, and push a change to a translation...
    - [x] in `securedrop/securedrop`
    - [x] in `securedrop/desktop`
7. [x] **In `securedrop-i18n-sandbox`,** review and merge the pull request.
8. [x] Test the release artifacts (no tags, signatures, or release objects are required):
    1. [x] `make dev` in `securedrop-i18n-sandbox@i18n` contains the current translations from `securedrop/securedrop`.
    2. [x] A staging server installed from packages built from `securedrop-i18n-sandbox` contains the current translations from `securedrop/securedrop`.
    4. [x] An Admin Workstation set up from `securedrop-i18n-sandbox` contains the current translations from `securedrop/desktop`.

Along the way, append `fixup` commits both here and in freedomofpress/securedrop-dev-docs#86 as needed for both a working implementation and correct documentation, respectively, of this workflow.
- [x] Rebase `fixup`s (only) before the final approval and merge.

## Deployment

Before merging:

1. [x] Give Localization Lab a heads-up.
4. [x] **At `weblate.securedrop.org`,** lock the `securedrop/securedrop` and `securedrop/desktop` components.
5. [x] Coordinate with Infrastructure so that:
    - [x] `weblate.securedrop.org` uses...
    - [x] GitHub user @weblate-fpf to open pull requests in...
    - [x] `freedomofpress/securedrop`.

After merging:

6. [x] **At `weblate.securedrop.org`,** configure `securedrop/securedrop` and `securedrop-desktop` (from `freedomofpress/securedrop`) like `securedrop/securedrop-client` (from `freedomofpress/securedrop-client`).  **Add-ons:** as above.
7. [x] **At `weblate.securedrop.org`,** unlock the `securedrop/securedrop` and `securedrop/desktop` components.
8. [x] Smoke-test a translation change.  It may take some cycling of the Weblate-side repository and add-ons before this yields as clean a diff as freedomofpress/securedrop-i18n-sandbox#3.  It should also pass CI into `securedrop`.

## Checklist

### If you made changes to the server application code:

- [x] Linting (`make lint`) and tests (`make test`) pass in the development container

### If you made non-trivial code changes:

- [x] I have written a test plan and validated it for this PR

Choose one of the following:

- [x] I have opened a PR in the [docs repo](https://github.com/freedomofpress/securedrop-docs) for these changes, or will do so later
    - freedomofpress/securedrop-dev-docs#86
- I would appreciate help with the documentation
- These changes do not require documentation

cfm committed:
chore: remove "i18n_tool.py"

cfm committed:
build({check,extract}-strings}): adapt from securedrop-client

cfm committed:
build(extract-strings): skip "securedrop/tests/"

cfm committed:
build(lint): run "check-strings"

This list was *almost* in alphabetical order, so I've reordered it.  If
there's a logical order in which we want these targets to run, we should
document that.

cfm committed:
build: inline "pybabel compile" invocation

As @legoktm notes[1] on a previous version of this commit, it's a single
line whether or not we wrap it in a Makefile target.

[1]: https://github.com/freedomofpress/securedrop/pull/6954#discussion_r1344693171

cfm committed:
build(supported-languages): read from "i18n.json"

cfm committed:
chore: "make extract-strings"

cfm committed:
chore: remove "list[-all]-translators" targets

cfm committed:
chore: types

cfm committed:
ci(translation-tests): split on output of "make supported-locales"

cfm committed:
test: don't use "i18n_tool.py"

Instead, generate test fixtures by calling "pybabel" directly.

cfm commented:
@nathandyer, I'd love your thoughts on this "test" plan, which is really a mock (compressed) QA/localization/release/deployment cycle.  Feel free to leave comments here, and we can discuss this (and next week's timeline) in Monday's stand-up too.

nathandyer commented:
@cfm I think this "test" plan makes a lot of sense, and is well thought out! I'm looking forward to pairing next week to work through this mock workflow together!

cfm commented:
@nathandyer, now that we've postponed the testing session to Thursday, would you be able to run through the async portion of the test plan beforehand?  (The very beginning of the "Test Plan" section, before the "Peform a Mock Release..." subsection.)  It would be nice to go into the end-to-end operational testing having already done a preliminary review of the self-contained functionality.

cfm committed:
build(check-supported-locales): render "i18n.json" to "i18n.rst"

Towards freedomofpress/securedrop-docs#503.

cfm committed:
chore: make "make" a little quieter and more portable

cfm committed:
ci(lint): run "check-supported-locales"

cfm committed:
ci(translation-tests): check that we have parallel coverage for all supported languages

legoktm commented on securedrop/i18n.py:
```diff
@@ -211,11 +211,11 @@ def get_locale(config: SecureDropConfig) -> str:
+    preferences: List[str] = []
     if session and session.get("locale"):
-        preferences.append(session.get("locale"))
+        preferences.append(str(session.get("locale")))
```
I would suggest a fix like https://github.com/freedomofpress/securedrop/commit/f334468fe89f670d2a96e61ba2dc0eb93e7c1c67#diff-805c9c6b4ac8e1734ec16c741576801472ffc790bb11c6e58f3506cc9de76002L214-R218

cfm replied:
Thanks!  Amended in 7418177354540df81a9fdb900fc5e12d33e78473.

legoktm commented on Makefile:
```diff
@@ -337,32 +340,79 @@ upgrade-destroy:  ## Destroy an upgrade test environment.
+
+.PHONY: compile-translations
+compile-translations: ## Compile translations for the Source and Journalist Interfaces.
+	@pybabel compile --directory ${LOCALE_DIR}
```
This command is basically as simple as the i18n_tool invocation it's replacing was - I would support inlining it in the 4 places its used. I personally find that clearer and easier to know what's going on instead of having to dig up the Makefile to find the underlying command being used.

cfm replied:
My instinct was to expose a consistent CLI (if you will).  But this is a fair point, and I've applied it in c7cc91d05a09e11a8b346236f2b03a419ba7412a.

legoktm commented on babel.cfg:
```diff
@@ -0,0 +1,13 @@
+# Extract from Python source files and Jinja templates.
```
Why do we need both this and `securedrop/babel.cfg`?

cfm replied:
Cognitive copypasta from `securedrop-client`, where this configuration lives in the root of the repository.  :-)  Deduplicated in 2659e8dd4a94263cb8575d2ae9c8db5c992c5bca.

legoktm reviewed (commented):
(very drive-by review)

nathandyer commented:
I've been trying to do the self-contained, local testing for this PR this morning, and every time I try it I run into the same error (with both this branch, and on `develop` itself).

I assume it's something I'm doing wrong, but with it all happening in containers, I can't quite put my finger on _how_. In multiple runs, on both branches, when I run `podman image prune -a --force && make build-debs` I receive this error (after it runs for several minutes):

```
FAILED builder/tests/test_securedrop_deb_package.py::test_app_code_paths[/opt/venvs/securedrop-app-code/lib/python3.8/site-packages/redwood/redwood.cpython-38-x86_64-linux-gnu.so] - Failed: not found
========================= 1 failed, 28 passed in 1.95s =========================
Script done.
```

Any ideas?

legoktm commented:
@nathandyer for develop, could you pastebin the build log and the output of `dpkg -c build/focal/securedrop-app-code_2.7.0~rc1+focal_amd64.deb` ?

cfm commented:
[`test_app_code_paths[.../redwood.cpython-38-x86_64-linux-gnu.so]` is passing in CI](https://app.circleci.com/pipelines/github/freedomofpress/securedrop/6360/workflows/cfc9812b-9f3f-47ef-9bda-099336a51a81/jobs/72271/steps?invite=true#step-102-2705176_178), for what it's worth.

zenmonkeykstop commented:
`make build-debs` failing on this branch for me too, build log attached
[20231005-securedrop.log](https://github.com/freedomofpress/securedrop/files/12820776/20231005-securedrop.log)

cfm commented:
NB.  A previous version of this branch made changes to `securedrop/debian/rules`—that should still have been independent of compilation and inclusion of the `redwood` bridge—but which would at least have suggested a locus for this breakage.  This version (after <https://github.com/freedomofpress/securedrop/pull/6954#discussion_r1344693171>) does not, which is why it's surprising.  @zenmonkeykstop, can you reproduce the same failure locally on `develop` too?

legoktm commented:
@zenmonkeykstop your failure seems to be a bit different because it couldn't set up the virtualenv? seems like a pre-existing issue in `devops/scripts/boot-strap-venv.sh` but not from this PR.

I did find an entirely different bug while testing this (issue coming in a moment), but unfortunately it wasn't Nathan's issue :/

nathandyer commented:
@legoktm Attached is the [build log](https://github.com/freedomofpress/securedrop/files/12821348/build.log) and [dpkg output](https://github.com/freedomofpress/securedrop/files/12821349/dpkg_output.txt) you asked for. Thanks!

legoktm commented:
After a bit of live debugging with Nathan, I've filed #6974, which is unrelated to this PR. Full speed ahead :)

zenmonkeykstop commented:
"You forgot to make clean" strikes again - it's working for me locally after nuking all containers and removing all non-repo artifacts.

nathandyer commented:
Thanks everyone! With those speed bumps out of the way, I'm happy to report that the local testing has been completed, and everything checks out. 

- Inspect packages for differences between `develop` and here, with a fresh build image each time:
    1. `(docker|podman) image prune -a --force && make build-debs` on `develop` and save them (e.g., `cp -Rp build build@develop`).
    2. `(docker|podman) image prune -a --force && make build-debs` on this branch.
    3. [x] `diffoscope build@develop/focal/securedrop-app-code*.deb build/focal/securedrop-app-code*.deb` shows only:
        - [x] diffs from this branch
        - [x] metadata changes in `.mo` files

cfm commented:
Thanks, @nathandyer!  Specifically, end-to-end testing succeeded for the `securedrop/securedrop` component, with the following fine print:

- We had to reset (force-pull, if you will) Weblate from `freedomofpress/securedrop-i18n-sandbox` (a) before it would pick up these changes and then (b) before it could push a clean pull request with a translation change.  I believe this to be a quirk of the sandbox configuration (no outgoing Web hook?) that we won't see in production, since I haven't seen it in `securedrop-client`.
- Even with line-wrapping disabled, there may still be some noisy diffs, especially in headers, until Weblate is consuming its own output in the _next_ localization cycle.

We skipped the `securedrop/desktop` component, because we realized that between #6953 and here I've neglected to script the few `desktop`-specific commands in this workflow.  I'll append that here later today.

That will pave the way for:
- [x] @nathandyer's final testing of the Admin/Journalist Workstation case on Tuesday;
- [x] coordination of approval and merge with freedomofpress/infrastructure#4393; and
- revision, approval, and merge of freedomofpress/securedrop-dev-docs#86 as part of (_not_ blocking) the v2.7.0 QA/localization/release process.

cfm committed:
build(update-supported-locales): also update desktop "LINGUAS" file

cfm committed:
build({check,update}-desktop-files): enforce msgfmt rendering of desktop files

cfm committed:
build(extract-strings): use Babel for unified extraction for securedrop/desktop too

cfm committed:
refactor(update-desktop-files): use "locale/$LANG/LC_MESSAGES/messages.po" structure

cfm committed:
docs: subsection "Localization" section for clarity

cfm commented:
@nathandyer, all the `securedrop/desktop` bits are wired up as of 0bd3249fefe8614d23ab7c1e128f78f420bf11a5 and demonstrated in freedomofpress/securedrop-i18n-sandbox@d6fcba4 into freedomofpress/securedrop-i18n-sandbox#3.  I'll take a look at the `lint` failure tomorrow to unblock your Tails-side testing on Tuesday.

cfm committed:
fix(extract-strings): sort securedrop/desktop template via "pybabel --sort-output"

I'm not sure what about the securedrop/desktop sources would otherwise
introduce sorting drift between my local environment and CI, but there's
no downside to applying this sorting wherever it turns out to be
necessary.

cfm commented:
Okay, we're out of the woods with 1735fa2354ee1f2126c5f9c3dbef18eea3420978.  @nathandyer, this was truly a lint-level fix, the gettext equivalent of re-running `black`.  So I won't bother pushing this to `securedrop-i18n-sandbox`, and you should be good to go on Tuesday morning per <https://github.com/freedomofpress/securedrop/pull/6954#issuecomment-1749883485>.

nathandyer commented:
Thank you @cfm!

I took this for a spin this morning, but it doesn't seem to be working as we might have hoped. The process I followed was:

1) I noticed where you had already modified a string in the extension template, changing `Launch Source Interface` to `Launch Interface`
2) In `weblate-sandbox.securedrop.org`, I found this change, and added a new translation for it under the French language
3) I committed the changes from the management view, then pushed them to the repository
4) In `securedrop-i18n-sandbox`, I reviewed the pull request,  confirmed it contained the necessary translation changes, and then merged it in
5) From an Admin workstation, I downloaded and extracted a copy of the `i18n` branch from `securedrop-i18n-sandbox`, then ran the usual `./securedrop-admin setup` and `./securedrop-admin --force tailsconfig`
6) I rebooted and confirmed the extension loaded (in English), and that it contained your string change
7) I rebooted and selected French (fr) as the language

Result:
After doing so, all the items in the extension are in French, except our new change; however, instead of the translation I provided, it just says "Launch Interface"

![Capture d’écran du 2023-10-10 16-05-12](https://github.com/freedomofpress/securedrop/assets/6131358/8bce4500-3ee0-4828-a565-f98ba47280ff)

cfm commented:
@nathandyer and I discussed <https://github.com/freedomofpress/securedrop/pull/6954#issuecomment-1755772800> out of band and concluded that it's mostly passing:

1. [x] <https://github.com/weblate-sandbox-fpf/securedrop-i18n-sandbox/blob/73b5aea826e6e6ec7a8f4ebe0f4c9994b183cce0/install_files/ansible-base/roles/tails-config/templates/locale/fr/LC_MESSAGES/messages.po#L21-L22> *is* shown in this screenshot.  The `msgid` `Launch Interface` (freedomofpress/securedrop-i18n-sandbox@d6fcba4) has an empty `msgstr`, so it's left untranslated, while all the other translations for `fr` are intact from Weblate.
2. [ ] <https://github.com/weblate-sandbox-fpf/securedrop-i18n-sandbox/blob/73b5aea826e6e6ec7a8f4ebe0f4c9994b183cce0/install_files/ansible-base/roles/tails-config/templates/locale/fr/LC_MESSAGES/messages.po#L18-L19> is correct too, but `make update-desktop-files` still has to be run manually as part of the release process:

    ```sh-session
    (.venv) user@sd-dev:~/securedrop-i18n-sandbox$ git whereami
    aa6495c2e (HEAD -> i18n, origin/i18n, origin/HEAD) Merge pull request #3 from weblate-sandbox-fpf/weblate-securedrop-securedrop
    (.venv) user@sd-dev:~/securedrop-i18n-sandbox$ make check-desktop-files
    make[2]: Nothing to be done for 'install_files/ansible-base/roles/tails-config/templates/ssh_config.j2'.
    Desktop files are out of date. Please run "make update-desktop-files" and commit the changes.
    make: *** [Makefile:395: check-desktop-files] Error 1
    ```

    This would have been [enforced](https://github.com/freedomofpress/securedrop/blob/1735fa2354ee1f2126c5f9c3dbef18eea3420978/Makefile#L143C1-L143C1) by `make lint` in CI, which isn't run for `securedrop-i18n-sandbox`.

    ```sh-session
    (.venv) user@sd-dev:~/securedrop-i18n-sandbox$ make update-desktop-files
    make[1]: Nothing to be done for 'install_files/ansible-base/roles/tails-config/templates/ssh_config.j2'.
    (.venv) user@sd-dev:~/securedrop-i18n-sandbox$ git diff
    diff --git a/install_files/ansible-base/roles/tails-config/templates/desktop-source-icon.j2 b/install_files/ansible-base/roles/tails-config/templates/desktop-source-icon.j2
    index 50efd29c2..13369da04 100644
    --- a/install_files/ansible-base/roles/tails-config/templates/desktop-source-icon.j2
    +++ b/install_files/ansible-base/roles/tails-config/templates/desktop-source-icon.j2
    @@ -6,27 +6,8 @@ Type=Application
     Terminal=false
     StartupNotify=true
     Categories=Network;
    -Name[ar]=واجهة SecureDrop للمصدر
    -Name[ca]=Interfície de font del SecureDrop
    -Name[cs]=SecureDrop rozhraní zdroje
    -Name[de_DE]=Quellenschnittstelle für SecureDrop
    -Name[el]=Διεπαφή πηγής του SecureDrop
    -Name[es_ES]=Interfaz de fuente de SecureDrop
    -Name[fr]=SecureDrop - Interface des sources
    -Name[hi]=SecureDrop स्रोत अंतराफलक
    -Name[is]=Heimildarmannaviðmót SecureDrop
    -Name[it]=Interfaccia Fonte di SecureDrop
    -Name[nb_NO]=Kildegrensesnitt for SecureDrop
    -Name[nl]=Securedrop - interface voor bronnen
    -Name[pt_BR]=Interface de Fontes do SecureDrop
    -Name[pt_PT]=Interface de Fontes do SecureDrop
    -Name[ro]=SecureDrop - interfața pentru surse
    -Name[ru]=SecureDrop — интерфейс источников
    -Name[sk]=SecureDrop rozhranie zdroja
    -Name[sv]=SecureDrop källgränssnitt
    -Name[tr]=SecureDrop Kaynak Arayüzü
    -Name[zh_Hans]=SecureDrop 线人界面
    -Name[zh_Hant]=SecureDrop 線人使用介面
    -Name=SecureDrop Source Interface
    +Name[es_ES]=missing a word
    +Name[fr]=Le SecureDrop Flufudo
    +Name=SecureDrop Interface
     Icon={{ tails_config_securedrop_dotfiles }}/securedrop_icon.png
     Exec=/usr/local/bin/tor-browser {{ item.0.onion_url }}
    ```

    All the subtractions are correct per freedomofpress/securedrop-i18n-sandbox@d6fcba4.  The additions are correct per freedomofpress/securedrop-i18n-sandbox#3.

    So, @nathandyer, as of freedomofpress/securedrop-i18n-sandbox@9b09f371a you should now see the Source Interface icon display the `msgstr` `Le SecureDrop Flufudo`.


I'll confirm this in a moment on my end.  I'll also rebase my own local copy of `securedrop-i18n-sandbox` after #6976 so that I can check the staging environment without rebasing and force-pushing here.

cfm commented:
<https://github.com/freedomofpress/securedrop/pull/6954#issuecomment-1756360296>:

> So, @nathandyer, as of freedomofpress/securedrop-i18n-sandbox@9b09f371a you should now see the Source Interface icon display the `msgstr` `Le SecureDrop Flufudo`.

Confirmed here:

* Under `en_US`: Source Interface icon is labeled `SecureDrop Interface`
* Under `fr_FR`:  Source Interface icon is labeled `Le SecureDrop Flufudo`

cfm commented:
> 2. [x] A staging server installed from packages built from `securedrop-i18n-sandbox` contains the current translations from `securedrop/securedrop`.

Confirmed that freedomofpress/securedrop-i18n-sandbox@9b09f371a98522f5120b60996f63c837460fa264 + #6975 includes <https://github.com/weblate-sandbox-fpf/securedrop-i18n-sandbox/blob/73b5aea826e6e6ec7a8f4ebe0f4c9994b183cce0/securedrop/translations/es_ES/LC_MESSAGES/messages.po#L866-L867>.

```sh-session
root@sd-staging:~/securedrop-i18n-sandbox# git show 55a37b831206bfbfdffe3b57babec678166f1cfb
commit 55a37b831206bfbfdffe3b57babec678166f1cfb (HEAD -> i18n)
Merge: 9b09f371a d431061c7
Author: root <root@(none)>
Date:   Tue Oct 10 23:13:34 2023 +0000

    Merge remote-tracking branch 'main/develop' into i18n
```

nathandyer reviewed (approved):
Everything is looking great to me, and I'm satisfied that our test plan (as unusual as it was, given that the author of the PR was directly involved in the testing itself) is sufficient and covers all the necessary bases. From my perspective, we're OK to go ahead and merge, then move this forward.

cfm commented:
Thanks, @nathandyer!  I'm about to merge and begin the fiddly bits between here and Weblate.

legoktm commented on Makefile:
```diff
@@ -138,7 +140,7 @@ yamllint:  ## Lint YAML files (does not validate syntax!).
 
 .PHONY: lint
-lint: ansible-config-lint check-ruff app-lint check-black html-lint shellcheck typelint yamllint ## Runs all lint checks
+lint: ansible-config-lint app-lint check-black check-desktop-files check-strings check-ruff check-supported-locales html-lint shellcheck typelint yamllint ## Runs all lint checks
```
Sorry I missed this, yes, there was slightly an order! `check-ruff` was second because it was the fastest thing and provided much quicker feedback than all the other tools.

cfm replied:
Fix proposed in #6995!

legoktm commented on securedrop/requirements/python3/translation-requirements.txt:
```diff
@@ -4,9 +4,14 @@
-babel==2.9.1 \
-    --hash=sha256:ab49e12b91d937cd11f0b67cb259a57ab4ad2b59ac7a3b41d6c06c0ac5b0def9 \
-    --hash=sha256:bc0c176f9f6a994582230df350aa6e05ba2ebe4b3ac317eab29d9be5d2768da0
+babel==2.12.1 \
```
Did this get a diff review?

cfm replied:
Good catch; thanks.  #6996 has the documentation fix; I'll do the diff reviews this afternoon.

cfm replied:
Done in <https://github.com/freedomofpress/securedrop-builder/wiki/Babel-2.9.1-to-2.12.1>.

legoktm commented on securedrop/requirements/python3/translation-requirements.txt:
```diff
@@ -4,9 +4,14 @@
+    # via
+    #   -r requirements/python3/translation-requirements.in
+    #   babelgladeextractor
+babelgladeextractor==0.7.0 \
```
This should also get a diff review since it's being installed/run in the build container.

cfm replied:
Done in <https://github.com/freedomofpress/securedrop-builder/wiki/BabelGladeExtractor-0.7.0>.
